File: update_activity.php
<?php
$content = file_get_contents('Activity_Diagrams_Rsix_Cell.drawio');

// Match Start cells
$content = preg_replace_callback('/(<mxCell[^>]+value="Start"[^>]+style="[^"]*ellipse[^"]*"[^>]*>)\s*<mxGeometry([^>]+)>/', function($matches) {
    $tag = $matches[1];
    $geom = $matches[2];
    
    // Replace value="Start" with value=""
    $tag = preg_replace('/value="Start"/', 'value=""', $tag);
    // Replace style="..." with new style
    $tag = preg_replace('/style="[^"]*"/', 'style="ellipse;html=1;fillColor=#000000;strokeColor=none;perimeter=ellipsePerimeter;"', $tag);
    
    // Parse geometry
    preg_match('/width="([^"]+)"/', $geom, $wm);
    preg_match('/height="([^"]+)"/', $geom, $hm);
    preg_match('/x="([^"]+)"/', $geom, $xm);
    preg_match('/y="([^"]+)"/', $geom, $ym);
    
    $w = isset($wm[1]) ? floatval($wm[1]) : 60;
    $h = isset($hm[1]) ? floatval($hm[1]) : 60;
    $x = isset($xm[1]) ? floatval($xm[1]) : 0;
    $y = isset($ym[1]) ? floatval($ym[1]) : 0;
    
    $nw = 30;
    $nh = 30;
    $nx = $x + ($w - $nw)/2;
    $ny = $y + ($h - $nh)/2;
    
    $geom = preg_replace('/width="[^"]+"/', 'width="' . $nw . '"', $geom);
    $geom = preg_replace('/height="[^"]+"/', 'height="' . $nh . '"', $geom);
    $geom = preg_replace('/x="[^"]+"/', 'x="' . $nx . '"', $geom);
    $geom = preg_replace('/y="[^"]+"/', 'y="' . $ny . '"', $geom);
    
    return $tag . "\n          <mxGeometry" . $geom . ">";
}, $content);

// Match End cells
$content = preg_replace_callback('/(<mxCell[^>]+value="End"[^>]+style="[^"]*ellipse[^"]*"[^>]*>)\s*<mxGeometry([^>]+)>/', function($matches) {
    $tag = $matches[1];
    $geom = $matches[2];
    
    $tag = preg_replace('/value="End"/', 'value=""', $tag);
    $tag = preg_replace('/style="[^"]*"/', 'style="ellipse;html=1;shape=endState;fillColor=#000000;strokeColor=#000000;"', $tag);
    
    preg_match('/width="([^"]+)"/', $geom, $wm);
    preg_match('/height="([^"]+)"/', $geom, $hm);
    preg_match('/x="([^"]+)"/', $geom, $xm);
    preg_match('/y="([^"]+)"/', $geom, $ym);
    
    $w = isset($wm[1]) ? floatval($wm[1]) : 60;
    $h = isset($hm[1]) ? floatval($hm[1]) : 60;
    $x = isset($xm[1]) ? floatval($xm[1]) : 0;
    $y = isset($ym[1]) ? floatval($ym[1]) : 0;
    
    $nw = 30;
    $nh = 30;
    $nx = $x + ($w - $nw)/2;
    $ny = $y + ($h - $nh)/2;
    
    $geom = preg_replace('/width="[^"]+"/', 'width="' . $nw . '"', $geom);
    $geom = preg_replace('/height="[^"]+"/', 'height="' . $nh . '"', $geom);
    $geom = preg_replace('/x="[^"]+"/', 'x="' . $nx . '"', $geom);
    $geom = preg_replace('/y="[^"]+"/', 'y="' . $ny . '"', $geom);
    
    return $tag . "\n          <mxGeometry" . $geom . ">";
}, $content);

file_put_contents('Activity_Diagrams_Rsix_Cell.drawio', $content);
echo "Done replacing Start and End elements.\n";

File: update_activity.py
import xml.etree.ElementTree as ET

def update_activity_diagram(filename):
    tree = ET.parse(filename)
    root = tree.getroot()
    
    # Process all mxCell elements
    for cell in root.findall('.//mxCell'):
        val = cell.get('value')
        style = cell.get('style', '')
        
        # Determine if this is a Start or End node
        is_start = False
        is_end = False
        
        if val == 'Start' and 'ellipse' in style:
            is_start = True
        elif val == 'End' and 'ellipse' in style:
            is_end = True
            
        if is_start or is_end:
            # Clear text
            cell.set('value', '')
            
            # Change style
            if is_start:
                cell.set('style', 'ellipse;html=1;fillColor=#000000;strokeColor=none;perimeter=ellipsePerimeter;')
            else:
                cell.set('style', 'ellipse;html=1;shape=endState;fillColor=#000000;strokeColor=#000000;')
                
            # Update geometry
            geom = cell.find('mxGeometry')
            if geom is not None:
                width = float(geom.get('width', '60'))
                height = float(geom.get('height', '60'))
                x = float(geom.get('x', '0'))
                y = float(geom.get('y', '0'))
                
                # We want new width=30, height=30
                # To keep it centered, shift x by (width - 30)/2
                new_width = 30.0
                new_height = 30.0
                
                dx = (width - new_width) / 2
                dy = (height - new_height) / 2
                
                new_x = x + dx
                new_y = y + dy
                
                # Format to int if possible, else round
                geom.set('width', str(int(new_width)))
                geom.set('height', str(int(new_height)))
                geom.set('x', str(int(new_x)))
                geom.set('y', str(int(new_y)))

    tree.write(filename, encoding='utf-8', xml_declaration=False)

if __name__ == '__main__':
    f = 'Activity_Diagrams_Rsix_Cell.drawio'
    update_activity_diagram(f)
    
    # Ensure <mxfile ...> header is preserved
    with open(f, 'r', encoding='utf-8') as file:
        content = file.read()
    if not content.startswith('<mxfile'):
        print("Warning: mxfile root tag might be malformed.")
